File: app/src/main/java/cc/smartcash/wallet/Fragments/TransactionFragment.java
package cc.smartcash.wallet.Fragments;

import android.os.Bundle;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.Spinner;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import cc.smartcash.wallet.Activities.MainActivity;
import cc.smartcash.wallet.Adapters.TransactionAdapter;
import cc.smartcash.wallet.Adapters.WalletSpinnerAdapter;
import cc.smartcash.wallet.Models.Transaction;
import cc.smartcash.wallet.Models.Wallet;
import cc.smartcash.wallet.R;
import cc.smartcash.wallet.Utils.SmartCashApplication;
import cc.smartcash.wallet.ViewModels.UserViewModel;

public class TransactionFragment extends Fragment {

    @BindView(R.id.all_transactions_underline)
    View allTransactionsUnderline;
    @BindView(R.id.received_underline)
    View receivedUnderline;
    @BindView(R.id.awaiting_underline)
    View awaitingUnderline;
    @BindView(R.id.paid_underline)
    View paidUnderline;
    View activeUnderline;
    @BindView(R.id.wallet_spinner)
    Spinner walletSpinner;
    SmartCashApplication smartCashApplication;
    private String activeFilter = null;
    private ArrayList<Wallet> walletList;
    private WalletSpinnerAdapter walletAdapter;
    private ArrayList<Transaction> filteredTransactions = new ArrayList<Transaction>();
    private ArrayList<Transaction> transactions;
    private TransactionAdapter transactionAdapter;

    public static TransactionFragment newInstance() {
        return new TransactionFragment();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        if (smartCashApplication == null)
            smartCashApplication = new SmartCashApplication(getContext());
        walletList = smartCashApplication.getUser(getContext()).getWallet();
        transactions = walletList.get(0).getTransactions();
        View view = inflater.inflate(R.layout.fragment_transaction, container, false);
        ButterKnife.bind(this, view);
        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        if (smartCashApplication == null)
            smartCashApplication = new SmartCashApplication(getContext());
        setupRecyclerViewTransactions();

        DisplayMetrics displayMetrics = this.getResources().getDisplayMetrics();
        walletSpinner.setDropDownWidth(displayMetrics.widthPixels);

        activeUnderline = getView().findViewById(R.id.all_transactions_underline);

        Spinner walletSpinner = getView().findViewById(R.id.wallet_spinner);
        walletAdapter = new WalletSpinnerAdapter(getContext(), walletList);
        walletSpinner.setAdapter(walletAdapter);

        walletSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                Wallet clickedItem = (Wallet) parent.getItemAtPosition(position);
                transactions = clickedItem.getTransactions();
                smartCashApplication.saveWallet(getContext(), walletList.get(position));
                setTransactions(activeFilter);
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {

            }
        });

        Wallet savedWallet = smartCashApplication.getWallet(getContext());

        if (savedWallet != null) {
            for (int i = 0; i < walletList.size(); i++) {
                if (savedWallet.getWalletId().equals(walletList.get(i).getWalletId())) {
                    walletSpinner.setSelection(i);
                }
            }
        }
    }

    private void setupRecyclerViewTransactions() {
        RecyclerView recyclerViewTransactions = getActivity().findViewById(R.id.transaction_recyclerview);
        LinearLayoutManager linearLayoutManagerTransactions = new LinearLayoutManager(getActivity(), RecyclerView.VERTICAL, false);
        transactionAdapter = new TransactionAdapter(getContext(), new ArrayList<>());

        recyclerViewTransactions.setLayoutManager(linearLayoutManagerTransactions);
        recyclerViewTransactions.setAdapter(transactionAdapter);

        transactionAdapter.setItems(transactions);
    }

    @OnClick({R.id.btn_all_transactions, R.id.btn_received, R.id.btn_awaiting, R.id.btn_paid, R.id.btn_att})
    public void onViewClicked(View view) {
        switch (view.getId()) {
            case R.id.btn_all_transactions:
                changeUnderline(allTransactionsUnderline);
                setTransactions(null);
                break;
            case R.id.btn_received:
                changeUnderline(receivedUnderline);
                setTransactions("Received");
                break;
            case R.id.btn_awaiting:
                changeUnderline(awaitingUnderline);
                setTransactions("Awaiting");
                break;
            case R.id.btn_paid:
                changeUnderline(paidUnderline);
                setTransactions("Sent");
                break;
            case R.id.btn_att:
                updateData();
                break;
        }
    }

    private void updateData() {
        UserViewModel model = ViewModelProviders.of(this).get(UserViewModel.class);

        model.getUser(smartCashApplication.getToken(getActivity()), getActivity()).observe(this, response -> {
            if (response != null) {
                Toast.makeText(getActivity(), "Updated!", Toast.LENGTH_LONG).show();
                smartCashApplication.saveUser(getActivity(), response);
                ((MainActivity) getActivity()).setWalletValue();
            } else {
                Log.e("Erro", "Erro ao buscar os valores!");
            }
        });
    }

    public void setTransactions(String filtro) {
        if (filtro == null)
            transactionAdapter.setItems(transactions);
        else {
            filteredTransactions = filterTransactions(filtro);
            transactionAdapter.setItems(filteredTransactions);
        }
        activeFilter = filtro;
    }

    public ArrayList<Transaction> filterTransactions(String filtro) {
        ArrayList<Transaction> filteredT = new ArrayList<>();

        for (Transaction item : transactions) {
            if (item.getDirection().equals(filtro))
                filteredT.add(item);
        }

        return filteredT;
    }

    public void changeUnderline(View newActiveUnderline) {
        activeUnderline.setVisibility(View.GONE);
        newActiveUnderline.setVisibility(View.VISIBLE);
        activeUnderline = newActiveUnderline;
    }
}

File: app/src/main/java/cc/smartcash/smarthub/Fragments/TransactionFragment.kt
package cc.smartcash.smarthub.Fragments

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import butterknife.BindView
import butterknife.ButterKnife
import butterknife.OnClick
import cc.smartcash.smarthub.Activities.MainActivity
import cc.smartcash.smarthub.Adapters.TransactionAdapter
import cc.smartcash.smarthub.Adapters.WalletSpinnerAdapter
import cc.smartcash.smarthub.Models.FullTransaction
import cc.smartcash.smarthub.Models.Wallet
import cc.smartcash.smarthub.R
import cc.smartcash.smarthub.Utils.SmartCashApplication
import cc.smartcash.smarthub.ViewModels.UserViewModel
import cc.smartcash.smarthub.tasks.TransactionTask
import cc.smartcash.smarthub.tasks.UpdateDataTask
import java.util.*

class TransactionFragment : Fragment() {

    private var smartCashApplication: SmartCashApplication? = null
    private var activeFilter: String? = null
    private var walletList: ArrayList<Wallet>? = null
    private var walletAdapter: WalletSpinnerAdapter? = null
    private var filteredTransactions = ArrayList<FullTransaction>()
    private var transactions: ArrayList<FullTransaction>? = null
    private var transactionAdapter: TransactionAdapter? = null

    @BindView(R.id.all_transactions_underline)
    lateinit var allTransactionsUnderline: View

    @BindView(R.id.received_underline)
    lateinit var receivedUnderline: View

    @BindView(R.id.awaiting_underline)
    lateinit var awaitingUnderline: View

    @BindView(R.id.paid_underline)
    lateinit var paidUnderline: View

    @BindView(R.id.fragment_receive_wallet_spinner)
    lateinit var walletSpinner: Spinner

    @BindView(R.id.login_main_loader)
    lateinit var loader: ProgressBar

    @BindView(R.id.frameLayout)
    lateinit var listTransactions: FrameLayout

    private var activeUnderline: View? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        if (smartCashApplication == null)
            smartCashApplication = SmartCashApplication(context!!)
        walletList = smartCashApplication!!.getUser()!!.wallet

        val view = inflater.inflate(R.layout.fragment_transaction, container, false)
        ButterKnife.bind(this, view)

        return view
    }


    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        if (smartCashApplication == null)
            smartCashApplication = SmartCashApplication(context!!)

        val displayMetrics = this.resources.displayMetrics
        walletSpinner.dropDownWidth = displayMetrics.widthPixels

        activeUnderline = getView()!!.findViewById(R.id.all_transactions_underline)
        walletSpinner = getView()!!.findViewById<Spinner>(R.id.fragment_receive_wallet_spinner)

         walletAdapter = WalletSpinnerAdapter(context!!, walletList!!)
        walletSpinner.adapter = walletAdapter

        val savedWallet = smartCashApplication!!.getWallet()

        if (savedWallet != null) {
            for (i in walletList!!.indices) {
                if (savedWallet.walletId == walletList!![i].walletId) {
                    walletSpinner.setSelection(i)
                }
            }
        }

        TransactionTask(context!!, ::showLoader, ::afterLoadTransactionsTask).execute()
    }

    private fun afterLoadTransactionsTask(transactionsResponse: ArrayList<FullTransaction>?) {
        transactions = transactionsResponse

        walletSpinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>, view: View, position: Int, id: Long) {
                val clickedItem = parent.getItemAtPosition(position) as Wallet
                transactions = clickedItem.transactions
                smartCashApplication!!.saveWallet(context!!, walletList!![position])
                setTransactions(activeFilter)
                setupRecyclerViewTransactions()
            }

            override fun onNothingSelected(parent: AdapterView<*>) {

            }
        }

        setupRecyclerViewTransactions()
        hiddenLoader()
    }

    private fun showLoader() {
        listTransactions.visibility = View.GONE
        loader.visibility = View.VISIBLE
    }

    private fun hiddenLoader() {
        loader.visibility = View.GONE
        listTransactions.visibility = View.VISIBLE
    }



    @OnClick(R.id.btn_all_transactions, R.id.btn_received, R.id.btn_awaiting, R.id.btn_paid, R.id.btn_att)
    fun onViewClicked(view: View) {
        when (view.id) {
            R.id.btn_all_transactions -> {
                changeUnderline(allTransactionsUnderline)
                setTransactions(null)
            }
            R.id.btn_received -> {
                changeUnderline(receivedUnderline)
                setTransactions("Received")
            }
            R.id.btn_awaiting -> {
                changeUnderline(awaitingUnderline)
                setTransactions("Awaiting")
            }
            R.id.btn_paid -> {
                changeUnderline(paidUnderline)
                setTransactions("Sent")
            }
            R.id.btn_att -> updateData()
        }
    }

    private fun setupRecyclerViewTransactions() {
        try{
            val recyclerViewTransactions = activity!!.findViewById<RecyclerView>(R.id.transaction_recyclerview)
            val linearLayoutManagerTransactions = LinearLayoutManager(activity, RecyclerView.VERTICAL, false)
            transactionAdapter = TransactionAdapter(context!!, ArrayList(), smartCashApplication!!.getWallet()!!.address!!)

            recyclerViewTransactions.layoutManager = linearLayoutManagerTransactions
            recyclerViewTransactions.adapter = transactionAdapter

            if(transactions!! != null)
                transactionAdapter!!.setItems(transactions!!)
        } catch (e: KotlinNullPointerException){
            android.widget.Toast.makeText(activity, "Error on Load Transactions, Try Login Again!", Toast.LENGTH_LONG).show()
        }
    }

    /*
    private fun updateData() {
        ViewModelProviders.of(this).get<UserViewModel>(UserViewModel::class.java).also {

            it.getUser(smartCashApplication!!.getToken()!!, activity!!).observe(this, androidx.lifecycle.Observer { response ->
                if (response != null) {
                    Toast.makeText(activity, getString(R.string.transaction_updated_message), Toast.LENGTH_LONG).show()

                    smartCashApplication!!.saveUser(activity!!, response)
                    (activity as MainActivity).setWalletValue()
                } else {
                    Log.e(TAG, getString(R.string.transaction_update_error_message))
                }
            })
        }
    }
     */

    private fun updateData() {
        UpdateDataTask(context!!, ::showLoader, ::afterUpdateTask).execute()
    }

    private fun afterUpdateTask() {
        Toast.makeText(activity, getString(R.string.send_message_success_return), Toast.LENGTH_LONG).show()
        hiddenLoader()
    }

    fun setTransactions(filtro: String?) {
        if (filtro == null)
            transactionAdapter!!.setItems(transactions!!)
        else {
            filteredTransactions = filterTransactions(filtro)
            transactionAdapter!!.setItems(filteredTransactions)
        }
        activeFilter = filtro
    }

    private fun filterTransactions(filtro: String): ArrayList<FullTransaction> {
        val filteredT = ArrayList<FullTransaction>()

        for (item in transactions!!) {
            if (FullTransaction.getDirection(item, smartCashApplication!!.getWallet()!!.address!!) == filtro)
                filteredT.add(item)
        }

        return filteredT
    }

    private fun changeUnderline(newActiveUnderline: View) {
        activeUnderline?.visibility = View.GONE
        newActiveUnderline.visibility = View.VISIBLE
        activeUnderline = newActiveUnderline
    }

    companion object {

        val TAG: String? = TransactionFragment::class.java.simpleName

        fun newInstance(): TransactionFragment {
            return TransactionFragment()
        }
    }
}
